### ISSUE TITLE
[P2] PPO 梯度累積的尾端未保存，正常結束再 resume 會遺失已計入 steps 的 batch

### ISSUE THREAD

**Issue by HHim8826** [HHim8826/Mortal#15]
審查版本：`f228ffffb7fab60115507ee68b32c1adba8d550e`。這是 `da1139c` 加入 --accumulate 後，`3c8475a` 的 checkpoint/resume 修正沒有涵蓋的新狀態。

[累積與 step：train_ppo.py:628–633](https://github.com/HHim8826/Mortal/blob/f228ffffb7fab60115507ee68b32c1adba8d550e/mortal/train_ppo.py#L628-L633)；[儲存欄位：438–457](https://github.com/HHim8826/Mortal/blob/f228ffffb7fab60115507ee68b32c1adba8d550e/mortal/train_ppo.py#L438-L457)；[重啟將 batches_in 歸零：464](https://github.com/HHim8826/Mortal/blob/f228ffffb7fab60115507ee68b32c1adba8d550e/mortal/train_ppo.py#L464)；[結束直接 save：747–752](https://github.com/HHim8826/Mortal/blob/f228ffffb7fab60115507ee68b32c1adba8d550e/mortal/train_ppo.py#L747-L752)。

每 --accumulate 個 microbatches 才執行 optimizer.step，但 round 邊界和 --rounds 的正常結束可能落在累積中途。checkpoint 雖保存 optimizer/scaler，這兩者都不包含 Parameter.grad；batches_in 也沒有保存。已讀的 batch 計入 steps、牌譜被消耗，但未更新的梯度在程序結束後消失。正常退出也會發生，不需要 crash。

已使用實際 main()、CPU 小型 Brain、相同兩個合成 batch，比較連續執行與中途正常結束後 resume，僅替換網路／輸入 loader／TensorBoard：
```text
--accumulate 2，每 round 1 batch

一次 --rounds 2：
  optimizer updates = 1
  saved steps = 2
  policy 參數最大改變 = 0.00030000135

兩次 --rounds 1，第二次 resume：
  optimizer updates = 0 + 0
  saved steps = 2
  policy 參數最大改變 = 0
```

預設 accumulate=64 時，一個 checkpoint 可以遺失最多 63 個已累積但未更新的 microbatches。分段執行每次都不足 64 時，steps/rounds 持續上升，卻可以永遠沒有一次更新。這是先前尾批／零進度問題的另一種形式。

建議：checkpoint 只能在完整 optimizer step 邊界建立，或保存並恢復 pending gradients、累積數與相應 scaler 狀態；正常結束也可明確 flush 尾端並依實際累積數校正縮放。以「連續跑」對「保存後 resume」的參數一致性作回歸驗證。

**Comment by HHim8826** [comment]
Fixed in a368665. `pending_grads` / `restore_pending`: the part-built gradient and `batches_in` are saved when a save lands off an accumulation boundary, and put back on resume. On a boundary nothing is written, since 43 MB of zeros is an expensive way to say there is nothing to carry.

Not done the way you listed first: checkpoints can still be taken mid-accumulation, and the tail is not flushed with a corrected scale at exit. Carrying the gradient keeps the step size exactly what a continuous run would have used, where a short flush would take a differently-weighted step -- that seemed the safer of the two for a run whose whole problem is a signal-to-noise ratio near 0.09.

A gradient accumulated over a different parameter set -- a trunk that was training then and is frozen now -- is dropped with a warning rather than loaded, because matching by position would put the trunk's gradient on the head.
